### PR TITLE
fix address sanitizer memory leak issues

### PR DESCRIPTION
**Pull request by jaysim**
this patch related with https://github.com/namjaejeon/ntfsprogs/issues/35

Found 56 memory leaks in mismatch_index_filename corrupted image. Fixed them.

I'll add more patchsets to resolve memory issues

### CODE REVIEW

File: src/ntfsck.c [hunk 1]
```diff
@@ -2343,6 +2365,8 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
 
 		ictx = ntfs_index_ctx_get(dir->ni, NTFS_INDEX_I30, 4);
 		if (!ictx) {
+			ntfs_attr_put_search_ctx(ctx);
+			goto next_dir;
```
Comment (namjaejeon) [inline src/ntfsck.c]: Ah,, You do newly add goto statement here. After this change, the below codes should be removed ?

File: src/ntfsck.c [hunk 2]
```diff
@@ -2357,6 +2381,9 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
 		if (ictx->block_size < NTFS_BLOCK_SIZE) {
 			ntfs_log_perror("Index block size (%d) is smaller than the "
 					"sector size (%d)", ictx->block_size, NTFS_BLOCK_SIZE);
+			goto next_dir;
+			ntfs_index_ctx_put(ictx);
```
Comment (namjaejeon) [inline src/ntfsck.c]: Ditto.

File: src/ntfsck.c [hunk 3]
```diff
@@ -2315,24 +2330,31 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
 		return 1;
 	}
 
+	if (ntfsck_check_directory(ni)) {
```
Comment (namjaejeon) [inline src/ntfsck.c:2403]: Hm.. Subject is memory leak. but there are many changes that are not related to the subject... in this patch..

File: src/ntfsck.c [hunk 4]
```diff
@@ -2385,6 +2414,9 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
 			ret = ntfsck_update_index_entry(ictx);
 			if (ret) {
 				fsck_err_failed();
+				goto next_dir;
```
Comment (namjaejeon) [inline src/ntfsck.c]: Ditto.

File: src/ntfsck.c [hunk 5]
```diff
@@ -2375,6 +2402,8 @@ static int ntfsck_scan_index_entries_btree(ntfs_volume *vol)
 			if (!ictx->ia_na) {
 				ntfs_log_perror("Failed to open index allocation of inode "
 						"%llu", (unsigned long long)dir->ni->mft_no);
+				goto next_dir;
```
Comment (namjaejeon) [inline src/ntfsck.c]: Ditto.

File: src/ntfsck.c (hunk 1 above)
Comment (jaysim) [reply src/ntfsck.c]: I fixed it.

File: src/ntfsck.c (hunk 3 above)
Comment (jaysim) [reply src/ntfsck.c:2403]: One of the root causes of memory leaks is repeatedly calling ntfs_inode_open() on an inode that has already been opened. This is particularly a problem with the root directory, which is opened first and then added to a list. When the list is traversed, the root directory, which is already open, is opened again. To resolve this issue, I first check the root directory and skip checking it (do not open it) during the loop of the list.

**Comment by jaysim** [comment]
fix for https://github.com/namjaejeon/ntfsprogs/issues/35#issuecomment-1427183234

**Comment by jaysim** [comment]
fix for https://github.com/namjaejeon/ntfsprogs/issues/35#issuecomment-1427207900

**Comment by jaysim** [comment]
fix for https://github.com/namjaejeon/ntfsprogs/issues/35#issuecomment-1427212307

File: src/ntfsck.c [hunk 6]
```diff
@@ -784,30 +788,24 @@ static void ntfsck_verify_mft_record(ntfs_volume *vol, s64 mft_num)
...
-
 			fsck_err_fixed();
 		}
 	}
-
-	/*
-	 * Update number of clusters that is used for each non-resident mft entries to
-	 * bitmap.
-	 */
-	ntfsck_update_lcn_bitmap(ni);
```
Comment (namjaejeon) [inline src/ntfsck.c:800]: when using no repair mode(-n) option, is there no issue ?

Comment (jaysim) [reply src/ntfsck.c:800]: Do you mean that ntfsck_update_lcn_bitmap() should be called in no repair mode?
If yes, could you please explain that?

I moved those function call to ntfsck_add_index_entry_orphaned_file(). If ntfsck_update_lcn_bitmap should be called in any mode, I'll dig it more to resolve it.

Comment (namjaejeon) [reply src/ntfsck.c:800]: If there is orphan files and run no repair mode fsck, lcn cluster bitmap check codes will find orphan clusters. because we don't set clusters bit of orphan fiels to fsck_lcn_bitmap.

Comment (jaysim) [reply src/ntfsck.c:800]: OK, I understand that even if fsck run with no repair mode, fsck should check current allocated bitmap to check whether orphan cluster or not. is right?
I'll check it and modify it if need.

Comment (jaysim) [reply src/ntfsck.c:800]: I modified updating lcn bitmap in https://github.com/namjaejeon/ntfsprogs/pull/36/commits/2e8a7884114cf4836031dbb037505e3e3c71cc38

**Comment by jaysim** [comment]
added fixing corrupted index block.
https://github.com/namjaejeon/ntfsprogs/pull/36/commits/25e77d3511e5556cbad65a754c6334b4aa8eb77a

**Comment by namjaejeon** [comment]
Applied 0000~0005 patches, I will apply 0006 patch after checking more.

**Comment by namjaejeon** [comment]
Applied the last patch. thanks.
